### PR TITLE
Fix/linting

### PR DESCRIPTION
**Pull request by merkata**
This PR fixes linting issues found by mypy and pylint linters that are run via `tox -e lint`.

### CODE REVIEW

File: tests/unit/__init__.py
Comment (arturo-seijas) [inline tests/unit/__init__.py:7]: You need to remove this. It's a breaking change in ops 2

File: pyproject.toml
Comment (arturo-seijas) [inline pyproject.toml:55]: Why this change? Is it possible to refactor to avoid it?

**Review by arturo-seijas — CHANGES_REQUESTED** [review]
I'm afraid you'll need to remove `ops.testing.CAN_CONNECT`instead of ignoring the linting error. This is the default behaviour in ops 2

File: pyproject.toml
Comment (merkata) [reply pyproject.toml:55]: Well it's a false-positive, pylint fails on this because there are several lines that are "found" in two files, the _patched_charm.py and test_charm.py, although in the first file the culprit is on line 61 with `SCRIPT_PATH = __import__("charm").SCRIPT_PATH` that seems to get expanded and pylint thinks it is a duplicate. I'm using this to reduce the noise. If there is another way I'd be happy to try, but this is what I ended up with.

File: tests/unit/__init__.py
Comment (merkata) [reply tests/unit/__init__.py:7]: Oh man, thanks for spotting! Will remove it.

File: tests/integration/helpers.py
Comment (arturo-seijas) [inline tests/integration/helpers.py:2]: Shouldn't we keep the copyright notice before the docs?

File: tests/integration/test_charm.py
Comment (arturo-seijas) [inline tests/integration/test_charm.py:2]: Same here

File: tests/integration/helpers.py
Comment (merkata) [reply tests/integration/helpers.py]: Good point, fixed!

File: tests/integration/test_charm.py
Comment (merkata) [reply tests/integration/test_charm.py]: Good point, fixed!

File: tests/integration/conftest.py
Comment (arturo-seijas) [inline tests/integration/conftest.py:1]: Same here

File: tests/unit/__init__.py
Comment (arturo-seijas) [inline tests/unit/__init__.py:1]: Same here

File: tests/unit/test_charm.py
Comment (arturo-seijas) [inline tests/unit/test_charm.py:2]: Why is this not a single line comment instead of a block?

Comment (merkata) [reply tests/unit/test_charm.py:2]: It's either me or `black`, but I bet it's me, will fix it :) Thanks!

File: tests/conftest.py
Comment (merkata) [reply tests/conftest.py]: Thanks, fixed!

File: tests/integration/conftest.py
Comment (merkata) [reply tests/integration/conftest.py]: Thanks, fixed!

File: tests/unit/__init__.py
Comment (merkata) [reply tests/unit/__init__.py]: Thanks, fixed!

**Review by arturo-seijas — DISMISSED** [review]
LGTM

File: pyproject.toml
Comment (jdkandersson) [reply pyproject.toml:55]: The alternative is to disable the rule specifically on the offending line, that way we retain the linting rule on the other parts of the repository

File: src/charm.py
Comment (jdkandersson) [inline src/charm.py:165]: This can be broken into multiple lines, e.g.:

```Python
"https://login.ubuntu.com/+saml": (
    "32:15:20:9F:A4:3C:8E:3E:8E:47:"
    "72:62:9A:86:8D:0E:E6:CF:45:D5"
)
```

File: src/charm.py
Comment (jdkandersson) [inline src/charm.py:320]: I'm guessing the offending line is the first one, should we instead summarise it and add a more detailed description in the body? E.g.;

```Python
        """Determine if the setup script is to be run.

        This is based on the current plan and the new S3 settings.

        Args:
            current_plan: Dictionary containing the current plan.
            s3info: S3Info object containing the S3 configuration options.

         Returns:
             If no services are planned yet (first run) or S3 settings have changed.
         """
```

File: src/charm.py
Comment (jdkandersson) [inline src/charm.py:464]: Same comment here, this can be broken into multiple lines. black will do this automatically if you just insert two quotes where you want the line break to be

File: tests/integration/conftest.py
Comment (jdkandersson) [inline tests/integration/conftest.py:80]: This can also be resolved using:

```Python
unit = ops_test.model.applications[app_name].units[0]
assert unit.workload_status == WaitingStatus.name
```

File: tests/integration/test_charm.py
Comment (jdkandersson) [inline tests/integration/test_charm.py:121]: For type: ignore, should we add a comment on why? This is requested by our contributing guide: https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#static-code-analysis

Comment (jdkandersson) [inline tests/integration/test_charm.py:121]: I think you can resolve this line to long as well using something like:

```Python
parsed_registration.body.find(
    "input", attrs={"name": "utf8"}
).get("value")  # type: ignore
```

Not sure where the type: ignore needs to go, just guessing

Comment (jdkandersson) [reply tests/integration/test_charm.py:121]: This comment also applies to other disabling of line too long in this PR

Comment (jdkandersson) [reply tests/integration/test_charm.py:121]: This also applies to any other type: ignore that were added

File: tests/unit/test_charm.py
Comment (jdkandersson) [inline tests/unit/test_charm.py:382]: If the `# pylint: disable=protected-access` is at the top of the file then it is disabled for the entire file. It is probably reasonable to do it for test files

File: tests/unit/test_charm.py
Comment (jdkandersson) [inline tests/unit/test_charm.py:393]: Should we note why this is reasonable to do?

**Review by jdkandersson — COMMENTED** [review]
Thanks for tackling the linting and typing issues!

File: src/charm.py
Comment (merkata) [reply src/charm.py:165]: Black puts it back on one line like  "32:15:20:9F:A4:3C:8E:3E:8E:47:" "72:62:9A:86:8D:0E:E6:CF:45:D5", but it is still less than the max-line-length, so it works. Thanks!

File: src/charm.py
Comment (merkata) [reply src/charm.py:320]: Definitely better, thanks!

File: tests/integration/conftest.py
Comment (merkata) [reply tests/integration/conftest.py:80]: Thanks, it is better that way!

File: tests/integration/test_charm.py
Comment (merkata) [reply tests/integration/test_charm.py:121]: I think we should address that, however I haven't authored the ignores, they were part of other PRs, perhaps on a subsequent refactor?

Comment (merkata) [reply tests/integration/test_charm.py:121]: Awesome, I tried something similar yesterday, but black reformatted it, however with your example it works! Thanks a lot!

File: tests/unit/test_charm.py
Comment (merkata) [reply tests/unit/test_charm.py:382]: Good point! I placed it on top and added a comment. Thanks!

File: tests/unit/test_charm.py
Comment (merkata) [reply tests/unit/test_charm.py:393]: Good point, noted!

**Review by arturo-seijas — DISMISSED** [review]
LGTM

File: tests/integration/test_charm.py
Comment (gregory-schiano) [inline tests/integration/test_charm.py:219]: We should probably keep the comment

File: tests/integration/test_charm.py
Comment (gregory-schiano) [inline tests/integration/test_charm.py:221]: Generic question, will black complain if we return before `.get(` ?
I don't find this kind of indentation very elegant and readable... but perhaps that's the python way

Comment (merkata) [reply tests/integration/test_charm.py:221]: Good question - unfortunately black reformatted it this way, I set it differently, though without luck. If it wasn't for the type comment, it would fit in one line as well, perhaps that's confusing black in this case?

File: tests/integration/test_charm.py
Comment (merkata) [reply tests/integration/test_charm.py:219]: We sure do, I don't know how it got deleted, I'm putting it back. Thanks!

**Review by arturo-seijas — APPROVED** [review]
LGTM

File: tests/integration/test_charm.py
Comment (gregory-schiano) [reply tests/integration/test_charm.py:221]: Well this suggests to use `\`
https://stackoverflow.com/questions/8683178/chained-method-calls-indentation-style-in-python
=> Which I like for readability

But PEP8 specifies:
https://peps.python.org/pep-0008/#maximum-line-length
`The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.`

Which would suggest this:
```python
csrf_token = (parsed_admin.head.find("meta", attrs={"name": "csrf-token"}) # type: ignore
        .get( "content"))
```
Which I don't find very nice

@jdkandersson ?

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py]: So if you just write it like this: 

`"./bin/bundle exec rake admin:create <<< "f"$'{email}\n{password}\n{password}\nY'"`

`black` should automatically break this up for you

File: tests/integration/test_charm.py
Comment (jdkandersson) [reply tests/integration/test_charm.py]: I'll leave that to you as a judgement call - it might be good to do it sooner rather than later to clean up the code base

File: tests/integration/test_charm.py
Comment (jdkandersson) [reply tests/integration/test_charm.py:221]: I read it differently, it implies what `black` is doing by breaking on the opening bracket of `.get(`, generally speaking on formatting - whatever black does is right 😉

Comment (jdkandersson) [reply tests/integration/test_charm.py:221]: note the `inside` in `Python’s implied line continuation inside parentheses`

Comment (jdkandersson) [reply tests/integration/test_charm.py:221]: Depending on which part the `# type: ignore` applies I think you can trick black into breaking somewhere else, although I usually do this through trial and error. E.g., you could try this, I'm not sure what black would do with it though:

```Python
csrf_token = parsed_admin.head.find(  # type: ignore
    "meta", attrs={"name": "csrf-token"}).get("content"
)
```

Comment (jdkandersson) [reply tests/integration/test_charm.py:221]: Anyway, the point of using black is so we don't need to debate formatting like this and just use whatever it does

Comment (merkata) [reply tests/integration/test_charm.py:221]: > Depending on which part the `# type: ignore` applies I think you can trick black into breaking somewhere else, although I usually do this through trial and error. E.g., you could try this, I'm not sure what black would do with it though:
> 
> ```python
> csrf_token = parsed_admin.head.find(  # type: ignore
>     "meta", attrs={"name": "csrf-token"}
> ).get("content")
> ```

Using that example, black re-formats it to the above (what is already commited). At this point I'm fine with black doing its thing. I don't have more suggestions.

Comment (gregory-schiano) [reply tests/integration/test_charm.py:221]: Well I don't find that black is doing something very readable in that case, having a single string and parenthesis on their own line look ugly to me.
```python
csrf_token = parsed_admin.head.find(  # type: ignore
    "meta", attrs={"name": "csrf-token"}
).get("content")
```

This formatting is way better and more readable. And for sure we can trust black to avoid formatting discussion, but these tools aren't the alpha and the omega... they're just using algorithm and could produce something that can be improved while still being valid :)

File: pyproject.toml
Comment (merkata) [reply pyproject.toml:55]: There is no offending line, placing the disable on some of the lines quoted in the error does not produce the desired result, even placing the disable on top of the file does not disable it, perhaps due to the import. More on the issue here https://github.com/PyCQA/pylint/issues/214 and also here https://github.com/snapcore/snapcraft/pull/3655

File: tests/integration/test_charm.py
Comment (merkata) [reply tests/integration/test_charm.py]: I'm seeing such ignores piling up in other repos as well, let's have a discussion on the topic, we should address the root of why `mypy` is complaining and how to fix it. I'm dropping it until we figure out a holistic approach for all of our charm repos.

File: tests/integration/test_charm.py
Comment (merkata) [reply tests/integration/test_charm.py:221]: Based on the feedback I don't know what to do next. I don't have approvals either, so is there a suggestion I can try and implement?

Comment (gregory-schiano) [reply tests/integration/test_charm.py:221]: I'm in favor of the formatting I quoted in the previous comment

Comment (merkata) [reply tests/integration/test_charm.py:221]: Yes, but

```shell
merkata@mastodon:~/dev/discourse-k8s-operator$ grep "csrf_token =" tests/integration/test_charm.py -A3
    csrf_token = parsed_admin.head.find(  # type: ignore
        "meta", attrs={"name": "csrf-token"}
    ).get("content")

merkata@mastodon:~/dev/discourse-k8s-operator$ black tests/integration/test_charm.py 
reformatted tests/integration/test_charm.py

All done! ✨ 🍰 ✨
1 file reformatted.
merkata@mastodon:~/dev/discourse-k8s-operator$ grep "csrf_token =" tests/integration/test_charm.py -A3
    csrf_token = parsed_admin.head.find("meta", attrs={"name": "csrf-token"}).get(  # type: ignore
        "content"
    )
```

**Review by amandahla — APPROVED** [review]
LGTM but it seems that trivy config was not found.

https://github.com/canonical/discourse-k8s-operator/actions/runs/4054111097/jobs/6975542344

Example of output in Indico:
```
"-a image" "-b table" "-c " "-d 1" "-e false" "-f os,library" "-g CRITICAL,HIGH" "-h " "-i ghcr.io/canonical/indico:4073362325" "-j ." "-k " "-l " "-m " "-n " "-o " "-p " "-q " "-r false" "-s " "-t " "-u " "-v trivy.yaml" "-z "
Running Trivy with trivy.yaml config from:  trivy.yaml
```

**Pull request merged**
